### ISSUE TITLE
setting class weights for multiclass classification

### ISSUE THREAD

**Issue by puzzlecollector** [dreamquark-ai/tabnet#377]
If I am training TabNet for multiclass classification with unbalanced labels, how do I set the class weights? 

for example, for lightgbm, I can set the class_weight = 'balanced' parameter when defining the model. 

<img width="867" alt="Screen Shot 2022-04-02 at 2 16 28 AM" src="https://user-images.githubusercontent.com/15623464/161311169-5b659102-a1ac-4c64-bc97-b994fe78ce78.png">

**Comment by Optimox** [comment]
As per the README.md:

`weights` : int or dict (default=0)

/!\ Only for TabNetClassifier Sampling parameter 
0 : no sampling
1 : automated sampling with inverse class occurrences
dict : keys are classes, values are weights for each class

**Comment by puzzlecollector** [comment]
@Optimox 

using sklearn's compute_class_weight I was able to calculate class weights for 3 classes 
<img width="765" alt="Screen Shot 2022-04-02 at 2 35 09 AM" src="https://user-images.githubusercontent.com/15623464/161313829-497a84ca-2797-4ea0-9655-571057bdcda4.png">

And then I create a dictionary with keys being the class types and values being the weights 

<img width="812" alt="Screen Shot 2022-04-02 at 2 35 43 AM" src="https://user-images.githubusercontent.com/15623464/161313920-6cf7ae1a-99c6-4d29-ad86-3e80627656f2.png">


Finally, I fit TabNet
<img width="667" alt="Screen Shot 2022-04-02 at 2 36 00 AM" src="https://user-images.githubusercontent.com/15623464/161313970-dd8be290-76c7-42da-a788-5487a547cc88.png">


But I end up with an error 
<img width="977" alt="Screen Shot 2022-04-02 at 2 36 13 AM" src="https://user-images.githubusercontent.com/15623464/161313991-f46f1480-506e-49a9-8e4c-304f34569966.png">

the error message says I should provide a list instead of a dictionary? can you help me resolve this issue? 😢

**Comment by puzzlecollector** [comment]
@Optimox 

I simply did this 

```
full_weights = [] 

for i in range(len(Y_train)):
    full_weights.append(d[Y_train[i][0]])
```

So basically for all the labels in the train data, checked if they were label 0,1 or 2 then assigned the calculated weights. And this seems to resolve the issue, although I am not so sure if this the correct approach.

**Comment by puzzlecollector** [comment]
@Optimox 

Also, I see a dramatic improvement in performance if I do not use balanced class weights (from test set accuracy of 63% to 74%). What could be the reason for this? am I using the class weights incorrectly? 

Interestingly, I tried classifying using lgbm too, and not using balanced class weights improves performance for lgbm as well (from 68% to 69%).

**Comment by Optimox** [comment]
You should be using TabNetClassifier and not TabNetMultiTaskClassifier as you only have one task, can you try to give weights as dictionnary with TabNetClassifier?

Accuracy is not a very good metric with unbalanced classes. If your most frequent class has 80% prevalance, a trivial model will have 80% accuracy. But a better model might end up with 75% accuracy but will predict all the classes with some mistakes 25% of the time. You should probably look at a different metric, maybe balanced accuracy or simply the logloss (or else).

**Comment by puzzlecollector** [comment]
@optimox 

Also is it expected for TabNetClassifier and TabNetMultiTaskClassifier to have different performances on the same data? Last time I tried on a different dataset, they seemed to have the same performance, but this time their performances are different, with TabNetMultiTaskClassifier having a superior performance than TabNetClassifier when both do not use class weights

**Comment by puzzlecollector** [comment]
@Optimox 

This is pure accuracy performance on test set (I will also check out other metrics like F1 and balanced accuracy) 

TabNetMultiTaskClassifier with class weights: 63% 
TabNetMultiTaskClassifier without class weights: 74% 
TabNetClassifier with class weights: 68% 
TabNetClassifier without class weights: 66% 

The dataset distribution is roughly 50% class 0, 35% class 1, 15% class 2. Class 2 occurs quite infrequently and it is actually quite insignificant, so it would be better if the model can predict class 0 and class 1 more precisely.

**Comment by Optimox** [comment]
For a binary classification problem, TabNetClassifier and TabNetMultiTaskClassifier whould be very similar (TabNetClassifier is just two times the same model). However if you have multiple classes, then each class is going to be consider independently in TabNetMultiTaskClassifier so you will have 3 different models for each class.

**Comment by puzzlecollector** [comment]
@Optimox ah I see. So it actually could be beneficial to use TabNetMultiTaskClassifier then? So using TabNetMultiTaskClassifier will train three TabNetClassifiers for each of the classes?

**Comment by Optimox** [comment]
Not exactly, you'll have one model with three separated heads for each task. The main difference is that the output probabilities won't sum to 1.

**Comment by Optimox** [comment]
Closing this, do not hesitate to reopen if you need further information.
